google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb

```diff
@@ -0,0 +1,566 @@
+              response = enum.next
+
+              case response.response_type
+              when :target_change
+                case response.target_change.target_change_type
+                when :NO_CHANGE
+                  # No change has occurred. Used only to send an updated
+                  # +resume_token+.
+
+                  @inventory.persist(
+                    response.target_change.resume_token,
+                    Convert.timestamp_to_time(
+                      response.target_change.read_time
+                    )
+                  )
+
+                  if @inventory.current? && @inventory.changes?
+                    synchronize do
+                      send_callback @inventory.build_query_snapshot
+                    end
+                  end
+                when :CURRENT
+                  # The targets reflect all changes committed before the targets
+                  # were added to the stream.
+                  #
+                  # This will be sent after or with a +read_time+ that is
+                  # greater than or equal to the time at which the targets were
+                  # added.
+                  #
+                  # Listeners can wait for this change if read-after-write
+                  # semantics are desired.
+
+                  @inventory.persist(
+                    response.target_change.resume_token,
+                    Convert.timestamp_to_time(
+                      response.target_change.read_time
+                    )
+                  )
+
+                  @inventory.current = true
+                when :RESET
+                  # The targets have been reset, and a new initial state for the
+                  # targets will be returned in subsequent changes.
+                  #
+                  # After the initial state is complete, +CURRENT+ will be
+                  # returned even if the target was previously indicated to be
+                  # +CURRENT+.
+
+                  raise RestartStream # Raise to restart the stream
+                end
+              when :document_change
+                # A {Google::Firestore::V1beta1::Document Document} has changed.
+
+                if response.document_change.removed_target_ids.any?
+                  @inventory.delete response.document_change.document.name
+                else
+                  @inventory.add response.document_change.document
+                end
+              when :document_delete
+                # A {Google::Firestore::V1beta1::Document Document} has been
+                # deleted.
+
+                @inventory.delete response.document_delete.document
+              when :document_remove
+                # A {Google::Firestore::V1beta1::Document Document} has been
+                # removed from a target (because it is no longer relevant to
+                # that target).
+
+                @inventory.delete response.document_remove.document
+              end
```

Comment:
This is missing the Existing Filter handling. See https://github.com/googleapis/nodejs-firestore/blob/master/src/watch.js#L762

Reply:
I don't see any documentation about this. What are we supposed to do for this? It looks like the node implementation does a full reset when the count is off. But the proto documentation says:

> If different from the count of documents in the client that match, the client must manually determine which documents no longer match the target.

How do we determine if the documents match the target or not?

Reply:
Okay, thanks. I will add this.

Bloom filters are... fun. Sorta.

Reply:
This has been added. Please take a look.

You will be happy to know that calculating the inventory count was the functionality that ultimately forced my hand to extract the listen functionality to a common object. `DocumentListener` and `QueryListener` still exist, because their callback mechanics and documentation are different, but they now use a common `Watch::Listener` object to do all the heavy lifting.